Edited text for clarity (#146)
canonical/ubuntu-server-documentation: explanation/intro-to/crypto-libraries.md

(introduction-to-crypto-libraries)=
# Introduction to cryptographic libraries 

> When choosing a crypto library, the following aspects should be considered to maintain security and compliance. Typically, you will want to answer questions such as: How can we ensure legacy crypto systems with known vulnerabilities are not being used? How can we enforce minimum key size requirements? And what criteria should we use to accept X.509 certificates when connecting to remote servers?
> However, the cryptographic library landscape is vast and complex, and there are many crypto libraries available on an Ubuntu system. When an application developer chooses a crypto library, they will consider many aspects, such as:

Reviewer: ```suggestion
When choosing a crypto library, the following aspects should be considered to maintain security and compliance. Typically, you will want to answer questions such as: How can we ensure legacy crypto systems with known vulnerabilities are not being used? How can we enforce minimum key size requirements? And what criteria should we use to accept X.509 certificates when connecting to remote servers?

However, the cryptographic library landscape is vast and complex, and there are many crypto libraries available on an Ubuntu system. When an application developer chooses a crypto library, they will consider many aspects, such as:
```

(introduction-to-crypto-libraries)=
# Introduction to cryptographic libraries

> System administrators often need to determine which cryptographic algorithms are being used on their systems because they require the answers to many crypto-related questions, including the following: How do they ensure no legacy crypto systems are being used? How do they make sure that no keys below a certain size are selected or created? And which types of X509 certificates are acceptable for connecting to remote servers?

Author: Works for me. Change made.

> However, the cryptographic library landscape is vast and complex, and there are many crypto libraries available on an Ubuntu system. When an application developer chooses a crypto library, they will consider many aspects, such as:

Author: Change made.

Each one of these has its own implementation details, API, behavior, configuration file, and syntax.

## Determining which libraries are being used by an application

Ultimately, the only reliable way to determine how an application uses cryptography is by reading its documentation or inspecting its source code. But even then, you may discover that the code is not available and sometimes the documentation might lack this information. If you are having problems finding answers to your crypto-related questions, there are some practical checks that can be made.

> Generally, the only way to find out these answers about crypto is for the sysadmin to go through a system and find out which crypto implementations are installed on the system and which configurations are being used. To make things even more complicated, sometimes an application implements its own crypto, without using anything external.

Reviewer: ```suggestion
To find out what libraries are being used, one generally needs to check all of the installed crypto implementations and their respective configurations. To make things even more complicated, sometimes an application implements its own crypto, without using anything external.

If you are having problems finding answers to your crypto-related questions, there are some practical checks that can be made.
```

Reviewer: I found the original sentence to be a bit circular and confusing - I think this is clearer

> The most common options are:

Reviewer: ```suggestion
```

Reviewer: This feels a bit redundant alongside all the other changes

 
### Check static linking

This is harder, as there is no dependency information in the binary package, and this usually requires inspection of the source package to see build dependencies. A detailed example of this approach is shown later in this page.

### Look at the plugins

The main binary of an application can not depend directly on a crypto library, but it could load dynamic plugins which do. Usually these would be packaged separately, and then we fall under the dynamic or static linking cases above. Note that via such a plugin mechanism, an application could depend on multiple external cryptographic libraries.

### Examine the execution of external binaries

The application could just plain call external binaries at runtime for its cryptographic operations, like calling out to `openssl` or `gnupg` to encrypt/decrypt data. This will hopefully be expressed in the dependencies of the package. If it's not, then it's a bug that should be reported.

### Indirect use of libraries or executables

The application could be using a third party library or executable which in turn could fall into any of the above categories.

### Check documentation more carefully

> Look at the application documentation again. It might have crypto options directly in its own configuration files, or point at specific crypto configuration files installed on the system. This may also provide you with assistance if the application uses external crypto libraries or whether it has its own implementation of crypto.

Reviewer: ```suggestion
Read the application documentation again. It might have crypto options directly in its own configuration files, or point at specific crypto configuration files installed on the system. This may also clarify whether the application uses external crypto libraries or it has its own implementation of crypto.
```

## Detailed examples of how to locate crypto libraries

### Use dpkg to check package dependencies

Since package dependencies are a good way to check what is needed at runtime by the application, you can find out which package owns what file with the `dpkg -S` command. For example:

```bash
$ dpkg -S /usr/bin/lynx
lynx: /usr/bin/lynx
```

> Now that you have the package name, check the package's dependencies and look for `Recommends`, as they are installed by default. Using the current example, we can now do the following:

Reviewer: ```suggestion
Now that you have the package name, check the package's dependencies. You should also look for `Recommends`, as they are installed by default. Using the current example, we can now do the following:
```

Each one of these has its own implementation details, API, behavior, configuration file, and syntax.

## Determining which libraries are being used by an application

Ultimately, the only reliable way to determine how an application uses cryptography is by reading its documentation or inspecting its source code. But even then, sysadmins may discover that the code is not available and sometimes the system documentation might lack this information. 

> Then the only way to find out these answers about crypto is for the sysadmin to go through a system and find out which crypto implementations are installed on the system and which configurations are being used. To make things even more complicated, sometimes an application implements its own crypto, without using anything external.

Author: Change made. Please check this since I am not sure I understood you.

> If you are seeking crypto-related answers to sysadmin-related questions and are having problems, there are some practical checks that can be made depending on how your application uses crypto:

Author: change made.

### Follow dynamic links

> This is the most common way, and very easy to spot via package dependencies and helper tools. A detailed example of this approach to determining crypto is discussed later in this page.

Author: The problem is that there are various approaches to spotting dependencies and not all of them are mentioned in this document to the same length. If we relocated the example content, it would be harder to read--in the original version, the list of all the approaches to identifying libraries is consistent. So what I have done is redone the heading for both samples of example content. The organization is generally the same, but now there is now a separate section for the detailed examples for locating static and dynamic libraries. 

In my new approach, the reader moves from a list of possible methods for identifying libraries to two detailed examples of how to locate libraries.

> This is the most common way, and very easy to spot via package dependencies and helper tools. A detailed example of this approach to determining crypto is discussed later in this page.

Reviewer: That seems like a sensible approach, thank you!

### Look at the plugins

The main binary of an application can not depend directly on a crypto library, but it could load dynamic plugins which do. Usually these would be packaged separately, and then we fall under the dynamic or static linking cases above. Note that via such a plugin mechanism, an application could depend on multiple external cryptographic libraries.

### Examine the execution of external binaries

The application could just plain call external binaries at runtime for its cryptographic operations, like calling out to `openssl` or `gnupg` to encrypt/decrypt data. This will hopefully be expressed in the dependencies of the package. If it's not, then it's a bug that should be reported.

### Indirect usage of libraries or executables

The application could be using a third party library or executable which in turn could fall into any of the above categories.

## Additional information for identifying crypto libraries

Here are some more specific tips for identifying the crypto libraries used by an application on an Ubuntu system:

### Check documentation more carefully

> Look at the application documentation again. It might have crypto options directly in its own configuration files, or point at specific crypto configuration files installed on the system. This may also provide you with assistance if the application uses external crypto libraries or if it has its own implementation of crypto.

Author: ok

### Use dpkg to check package dependencies (with example)

> Since package dependencies are a good way to check what is needed at runtime by the application, you can find out which package owns what file with the `dpkg -S` command. For example:

Reviewer: I like this compaction :)

```bash
$ dpkg -S /usr/bin/lynx
lynx: /usr/bin/lynx
```

> Now that you have the package's name in hand, check the package's dependencies and look for `Recommends`, as they are installed by default. Continuing with the current example, we can now do the following:

Reviewer: ```suggestion
Now you have the package name, check the package dependencies and look for `Recommends`, as they are installed by default. Using the current example, we can do this by running:
```

Author: Typically, I avoid editing instructions when I don't know the context. Thanks.

```bash
$ dpkg -s lynx | grep -E "^(Depends|Recommends)"
Depends: libbsd0 (>= 0.0), libbz2-1.0, libc6 (>= 2.34), libgnutls30 (>= 3.7.0), libidn2-0 (>= 2.0.0), libncursesw6 (>= 6), libtinfo6 (>= 6), zlib1g (>= 1:1.1.4), lynx-common
Recommends: mime-support
```

Now we can see that `lynx` links with `libgnutls30`, which answers our question: `lynx` uses the GnuTLS library for its cryptography operations.

### Using ldd to list dynamic libraries (with example)

> If an dynamic library is needed by an application, it should always be correctly identified in the list of dependencies for the application package. When that is not the case, or you need to identify what is needed by some plugin that is not part of the package, you can use some system tools to help identify the dependencies.

Reviewer: ```suggestion
If a dynamic library is needed by an application, it should always be correctly identified in the list of application package dependencies. When that is not the case, or you need to identify what is needed by some plugin that is not part of the package, you can use some system tools to help identify the dependencies.
```

Reviewer: Spotted an opportunity for further compaction :)

Author: change made.

> In this situation, you can employ `ldd`, which is installed in all Ubuntu systems and will list all the dynamic libraries that are needed by the given binary, including dependencies of dependencies, i.e. the command is recursive. Going back to the `lynx` example:

Reviewer: ```suggestion
In this situation, you can use the `ldd` tool, which is installed in all Ubuntu systems. It lists all the dynamic libraries needed by the given binary, including dependencies of dependencies, i.e. the command is recursive. Going back to the `lynx` example:
```

Reviewer: The suggested change makes the sentence a bit too long.

We again see the GnuTLS library (via `libgnutls.so.30`) in the list, and can reach the same conclusion.

> Another way to check for such dependencies but without recursion, is via `objdump`. You may need to be install it with the `binutils` package, as it's not mandatory.

Reviewer: ```suggestion
Another way to check for such dependencies (without recursion) is via `objdump`. You may need to install it with the `binutils` package, as it's not mandatory.
```

Author: change made.

    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6
```

> ### Check package headers for static linking (with example)

Reviewer: ```suggestion
### Check package headers for static linking
```

Reviewer: This is already in the examples section, so we can safely remove this :)

    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6
```

### Check package headers for static linking (with example)

Identifying which libraries were used in a static build is a bit more involved. There are two ways, and they are complementary most of the time:

* look for the `Built-Using` header in the binary package
* inspect the `Build-Depends` header in the source package

> For example, let's try to discover which crypto libraries, if any, the `rclone` tool uses for an example application. First, let's try the packaging dependencies:

Reviewer: This change adds repetition, as the start of the sentence already calls this an example. Could we revert this change?

Author: change reverted.

```bash
$ dpkg -s rclone | grep -E "^(Depends|Recommends)"
Depends: libc6 (>= 2.34)
```

> Yes, that's a short list. But `rclone` definitely supports encryption, so what is going on? Turns out this is a tool written in the Go language, and that uses static linking of libraries. So let's try to inspect the package data more carefully, and this time look for the `Built-Using` header:

Reviewer: I'd also like to revert this change too, please. I left this in deliberately as it was an expression of the original author's "voice" - where possible, I like to leave those alone.

Author: change reverted.
